api/envoy/extensions/filters/http/jwt_authn/v3/config.proto

```diff
@@ -119,6 +119,35 @@ message JwtProvider {
   //
   type.matcher.v3.StringMatcher subjects = 19;
 
+  oneof lifetime_constraint {
+    // Restrict the maximum remaining lifetime of a credential from the JwtProvider. Credential lifetime
+    // is the difference between the current time and the expiration of the credential. For instance,
+    // the following example will reject credentials that have a lifetime longer than 24 hours. If not set,
+    // expiration checking still occurs, but there is no limit on credential lifetime.
+    //
+    // Example:
+    //
+    // .. code-block:: yaml
+    //
+    //     max_lifetime:
+    //       seconds: 86400
+    //
+    google.protobuf.Duration max_lifetime = 20;
+
+    // Requires that the credential contains an `expiration <https://tools.ietf.org/html/rfc7519#section-4.1.4>`_.
+    // For instance, this could implement JWT-SVID
+    // `expiration restrictions <https://github.com/spiffe/spiffe/blob/main/standards/JWT-SVID.md#33-expiration-time>`_.
+    // Unlike ``max_lifetime``, this only requires that expiration is present, where ``max_lifetime`` also checks the value.
+    //
+    // Example:
+    //
+    // .. code-block:: yaml
+    //
+    //     require_expiration: true
+    //
+    bool require_expiration = 21;
```

Comment:
Thanks, this makes it much clearer!  In general, the new guidance is to avoid `oneof`, see https://github.com/envoyproxy/envoy/blob/3c84a245c056ae5bd74309d766091f33d9190071/api/STYLE.md?plain=1#L63.

If you want to require one or the other (either max_lifetime must be set, or require_expiration must be set), why not enforce that in the code?  I.e. just have `max_lifetime` field... if it is not set, the code checks that the JWT has an expiration on it; if not, raise an error condition.

Reply:
Thanks for the link, very helpful. There's a 3rd case where neither is set because the JWT spec does not require expiration for a token. And, envoy does not currently require one, so setting nothing and implying require_expiration would change behavior for existing deployments. But, based on the guide I think this is very similar to the example where max_lifetime is a superset of require_expiration. We could also have some special value in the duration, like setting duration to 0 for require any expiration, but I think that's not as clear. Let me try one more.